### ISSUE TITLE
DataStore.query in Next.js getServerSideProps prevents rendering page

### ISSUE THREAD

**Issue by MarceloAlves** [aws-amplify/amplify-js#7412]
**Describe the bug**

Currently I have an Amplify app that was set up in the Admin UI with one model and using Cognito user pool for Authentication. When trying to access the data in NextJS using `getServerSideProps()` the page won't render. Loading the data outside of server side props works just fine.

I originally thought it could be an issue with how the project is set up so I deleted the app, set up a new one, pulled changes and tried again.

**To Reproduce**

1. Start new Next.js Project
2. Create Amplify project from UI
3. Create model in Admin UI
4. Set up authentication
5. Pull changes into app
6. Try to call `DataStore.query()` from `getServerSideProps`

**Expected behavior**

Should return data and render page

**Code Snippet**

```tsx
import { serializeModel } from '@aws-amplify/datastore/ssr'
import { Amplify, withSSRContext } from 'aws-amplify'
import awsExports from '../../aws-exports'
import { Cook } from '@models/index'

Amplify.configure({ ...awsExports, ssr: true })

export default function Cooks() {
  return <div>Index</div>
}

export async function getServerSideProps({ req }: any) {
  const { DataStore } = withSSRContext({ req })
  const cooks = await DataStore.query(Cook)

  return {
    props: {
      cooks: serializeModel(cooks),
    },
  }
```

After some minutes an error appears in the console:

```
event - build page: /cooks
wait  - compiling...
event - compiled successfully
The schema has already been initialized
[WARN] 28:47.268 DataStore - Realtime disabled when in a server-side environment
(node:38173) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'some' of undefined
    at SyncProcessor.<anonymous> (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:154:71)
    at step (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:33:23)
    at Object.throw (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:14:53)
    at rejected (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:6:65)
(node:38173) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:38173) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:38173) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'some' of undefined
    at SyncProcessor.<anonymous> (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:154:71)
    at step (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:33:23)
    at Object.throw (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:14:53)
    at rejected (/Users/marceloalves/Development/marceloalves/sampleapp/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:6:65)
(node:38173) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
```

**Screenshots**

N/A

**What is Configured?**
If applicable, please provide what is configured for Amplify CLI:
* Which steps did you follow via Amplify CLI when configuring your resources.
  
  Used `amplify pull --appId=xxxx -envName=staging` to bring in changes

* Which resources do you have configured?
  * If applicable, please provide your `aws-exports` file:
  ```
  const awsmobile = {
      "aws_project_region": "us-west-2",
      "aws_cognito_identity_pool_id": "xxxx",
      "aws_cognito_region": "us-west-2",
      "aws_user_pools_id": "xxxx",
      "aws_user_pools_web_client_id": "xxxx",
      "oauth": {},
      "aws_appsync_graphqlEndpoint": "xxxx",
      "aws_appsync_region": "us-west-2",
      "aws_appsync_authenticationType": "AMAZON_COGNITO_USER_POOLS"
  };
  ```

* If applicable, provide more configuration data, for example for Amazon Cognito, run `aws cognito-idp describe-user-pool --user-pool-id us-west-2_xxxxxx` (Be sure to remove any sensitive data)

```
{
    "UserPool": {
        "Id": "xxxx",
        "Name": "sampleapp-staging",
        "Policies": {
            "PasswordPolicy": {
                "MinimumLength": 8,
                "RequireUppercase": true,
                "RequireLowercase": true,
                "RequireNumbers": true,
                "RequireSymbols": true,
                "TemporaryPasswordValidityDays": 7
            }
        },
        "LambdaConfig": {},
        "LastModifiedDate": 1608170131.833,
        "CreationDate": 1608170131.833,
        "SchemaAttributes": [
            {
                "Name": "sub",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": false,
                "Required": true,
                "StringAttributeConstraints": {
                    "MinLength": "1",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "name",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": true,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "given_name",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "family_name",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "middle_name",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "nickname",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "preferred_username",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": true,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "profile",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "picture",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "website",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "email",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": true,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "email_verified",
                "AttributeDataType": "Boolean",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false
            },
            {
                "Name": "gender",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "birthdate",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "10",
                    "MaxLength": "10"
                }
            },
            {
                "Name": "zoneinfo",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "locale",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "phone_number",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": true,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "phone_number_verified",
                "AttributeDataType": "Boolean",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false
            },
            {
                "Name": "address",
                "AttributeDataType": "String",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "StringAttributeConstraints": {
                    "MinLength": "0",
                    "MaxLength": "2048"
                }
            },
            {
                "Name": "updated_at",
                "AttributeDataType": "Number",
                "DeveloperOnlyAttribute": false,
                "Mutable": true,
                "Required": false,
                "NumberAttributeConstraints": {
                    "MinValue": "0"
                }
            }
        ],
        "AutoVerifiedAttributes": [
            "email"
        ],
        "SmsVerificationMessage": "Your verification code is {####}",
        "EmailVerificationMessage": "Forgot password code: {####}",
        "EmailVerificationSubject": "Forgot password code: {####}",
        "VerificationMessageTemplate": {
            "SmsMessage": "Your verification code is {####}",
            "EmailMessage": "Forgot password code: {####}",
            "EmailSubject": "Forgot password code: {####}",
            "DefaultEmailOption": "CONFIRM_WITH_CODE"
        },
        "MfaConfiguration": "OPTIONAL",
        "EstimatedNumberOfUsers": 1,
        "EmailConfiguration": {
            "EmailSendingAccount": "COGNITO_DEFAULT"
        },
        "SmsConfiguration": {
            "SnsCallerArn": "xxxx",
            "ExternalId": "xxxx"
        },
        "UserPoolTags": {},
        "AdminCreateUserConfig": {
            "AllowAdminCreateUserOnly": false,
            "UnusedAccountValidityDays": 7
        },
        "Arn": "xxxx"
    }
}
```
 

<details>
  <summary><strong>Environment</strong></summary>

```
npx envinfo --system --binaries --browsers --npmPackages --npmGlobalPackages

  System:
    OS: macOS 11.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 2.53 GB / 32.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 12.20.0 - /usr/local/bin/node
    Yarn: 1.22.5 - /usr/local/bin/yarn
    npm: 6.14.8 - /usr/local/bin/npm
  Browsers:
    Chrome: 87.0.4280.88
    Edge: 87.0.664.60
    Firefox: 77.0.1
    Firefox Developer Edition: 84.0
    Safari: 14.0.2
  npmPackages:
    @aws-amplify/ui-react: ^0.2.32 => 0.2.32
    @chakra-ui/react: ^1.0.3 => 1.0.3
    @emotion/react: ^11.1.2 => 11.1.2
    @emotion/styled: ^11.0.0 => 11.0.0
    @hookform/devtools: ^2.2.1 => 2.2.1
    @hookform/resolvers: ^1.2.0 => 1.2.0
    @types/node: ^14.14.14 => 14.14.14
    @types/react: ^17.0.0 => 17.0.0
    @typescript-eslint/eslint-plugin: ^4.10.0 => 4.10.0
    @typescript-eslint/parser: ^4.10.0 => 4.10.0
    aws-amplify: ^3.3.12 => 3.3.12
    eslint: ^7.15.0 => 7.15.0
    eslint-config-prettier: ^7.0.0 => 7.0.0
    eslint-loader: ^4.0.2 => 4.0.2
    eslint-plugin-jsx-a11y: ^6.4.1 => 6.4.1
    eslint-plugin-prettier: ^3.3.0 => 3.3.0
    eslint-plugin-react: ^7.21.5 => 7.21.5
    eslint-plugin-react-hooks: ^4.2.0 => 4.2.0
    framer-motion: ^2.9.5 => 2.9.5
    husky: ^4.3.6 => 4.3.6
    next: 10.0.3 => 10.0.3
    prettier: ^2.2.1 => 2.2.1
    pretty-quick: ^3.1.0 => 3.1.0
    react: 17.0.1 => 17.0.1
    react-dom: 17.0.1 => 17.0.1
    react-hook-form: ^6.13.1 => 6.13.1
    react-query: ^3.2.0 => 3.2.0
    typescript: ^4.1.3 => 4.1.3
    yup: ^0.32.8 => 0.32.8
  npmGlobalPackages:
    @aws-amplify/cli: 4.20.0
    alfred-vscode: 2.1.0
    caniuse-lite: 1.0.30001036
    dts-gen: 0.6.0
    jscodeshift: 0.7.0
    local-web-server: 4.2.1
    npm: 6.14.8
    react-query: 1.0.7
    react: 16.13.1
    svgexport: 0.4.0
    swaxios: 0.1.4
```

</details>

**Smartphone (please complete the following information):**

N/A

**Additional context**

N/A

**_You can turn on the debug mode to provide more info for us by setting window.LOG_LEVEL = 'DEBUG'; in your app._**

**Comment by tabzone** [comment]
I am having the same issue .

**Comment by dbhagen** [comment]
Same here.
```javascript
(node:75647) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'some' of undefined
    at SyncProcessor.<anonymous> (/Users/dhagen/Projects/****/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:154:71)
    at step (/Users/dhagen/Projects/****/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:33:23)
    at Object.throw (/Users/dhagen/Projects/****/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:14:53)
    at rejected (/Users/dhagen/Projects/****/node_modules/@aws-amplify/datastore/lib/sync/processors/sync.js:6:65)
(node:75647) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 21)
```
I'm on the latest version of `@aws-amplify/datastore`, `2.9.3` and `aws-amplify` is `3.3.14`
Amplify CLI is `4.41.1`

My call right now is:
```typescript
import { withSSRContext } from 'aws-amplify'
import { PartnerOAuth } from 'src/models'

export const getServerSideProps: GetServerSideProps = async ({ req, res }) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { Auth, DataStore } = withSSRContext({ req })

await DataStore.save(
    new PartnerOAuth({
      partner: 'YNAB',
      token: '123',
      misc: JSON.stringify({ Thing: 'Data' }),
    })

...
}
```

**Comment by dbhagen** [comment]
If I had to theorize, I'm thinking that SSR is trying to access the API but doesn't have an authenticated Cognito user to allow this call? And even if I've got an API key, that key might not be authorized for the query being attempted.

I'm going to run a quick test to see if removing some of the auth aspects unblocks this.
@ericclemmons any other thoughts?

**Comment by medelman17** [comment]
I ran into this (or essentially the same) issue today. For me, (1) everything works as expected initially; (2) updates to schema where Dynamo has existing data cause DataStore operations to timeout/fail silently; (3) purging/dropping data from Dynamo tables seems to get everything working again.

 @dbhagen, do you have any further insight based on your tests outlined above?

**Comment by ericclemmons** [comment]
This is an interesting bug, because `.some` _is_ gated by multiple checks to prevent `UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'some' of undefined`:

https://github.com/aws-amplify/amplify-js/blob/a5f2466454f0cc1e0223fb03e78142209cf1f033/packages/datastore/src/sync/processors/sync.ts#L248-L253

**Comment by amhinson** [comment]
@MarceloAlves @dbhagen Are you still experiencing this issue with the latest version of `aws-amplify`? There have been some updates around authorization in DataStore, so it is possible that the `UnhandledPromiseRejectionWarning` error no longer occurs.

**Comment by MarceloAlves** [comment]
Hi @amhinson 

Like most sideprojects, I have tried this in a long while. Maybe I can start up a new project and give it a shot

**Comment by PhilipWee** [comment]
Hey guys, I'm having a similar problem, I think that Amplify just works differently server side, you can view the following docs to have a better idea:

https://docs.amplify.aws/lib/ssr/q/platform/js

TLDR:
`import { Amplify } from "aws-amplify";
import awsExports from "../src/aws-exports";

Amplify.configure({ ...awsExports, ssr: true });`

**Comment by JosXa** [comment]
We can reproduce this. It appears that the first `.query` just hangs seemingly forever, this being in NextJS `App.getInitialProps` however. It seems like this has something to do with model changes, as it worked previously and then just started happening after running a schema migration in Amplify Studio.
Any debug information that could be enabled to help you triage this?
